cmd/xl-v1-metadata.go

```diff
@@ -17,20 +17,92 @@
 package cmd
 
 import (
+	"crypto"
+	"crypto/sha256"
+	"encoding/hex"
 	"encoding/json"
 	"errors"
+	"fmt"
+	"hash"
 	"path"
-	"runtime"
 	"sort"
 	"sync"
 	"time"
+
+	"golang.org/x/crypto/blake2b"
 )
 
+const erasureAlgorithmKlauspost = "klauspost/reedsolomon/vandermonde"
+
+// DefaultBitrotAlgorithm is the default algorithm used for bitrot protection.
+const DefaultBitrotAlgorithm = BLAKE2b512
+
+func init() {
+	newBLAKE2b := func() hash.Hash {
+		b2, _ := blake2b.New512(nil) // New512 never returns an error if the key is nil
+		return b2
+	}
+	crypto.RegisterHash(crypto.Hash(SHA256), sha256.New)
+	crypto.RegisterHash(crypto.Hash(BLAKE2b512), newBLAKE2b)
+	crypto.RegisterHash(crypto.Hash(HighwayHash256), nil) // TODO(aead): currently not supported, waiting for google to finish algorithm spec.
+}
+
+// BitrotAlgorithm specifies a algorithm used for bitrot protection.
+type BitrotAlgorithm crypto.Hash
+
 const (
-	// Erasure related constants.
-	erasureAlgorithmKlauspost = "klauspost/reedsolomon/vandermonde"
+	// SHA256 represents the SHA-256 hash function
+	SHA256 = BitrotAlgorithm(crypto.SHA256)
+
+	// HighwayHash256 represents the HighwayHash-256 hash function
+	HighwayHash256 = BitrotAlgorithm(crypto.SHA3_256) // we must define that HighwayHash-256 is SHA3-256 because there is no HighwayHash constant in golang/crypto yet.
+
+	// BLAKE2b512 represents the BLAKE2b-256 hash function
+	BLAKE2b512 = BitrotAlgorithm(crypto.SHA3_512) // we must define that BLAKE2b-512 is SHA3-512 because there is no BLAKE2b-512 constant in golang/crypto yet (Go 1.9).
```

Comment:
It's a temp solution. (namespace separation) - BLAKE2 constants are already in Go1.9